File: 2022/09/Javascript/main.js
const Direction = {
	"U": 1,  // up
	"D": 2,  // down
	"L": 4,  // left
	"R": 8,  // right
};

class Knot {

	constructor ( row, col ) {
		this.row = row;
		this.col = col;
		this.history = {};
		this.track_history();
	}

	track_history () {
		let key = this.row + "," + this.col;
		this.history[key]++;
	}

	get history_size () {
		return Object.keys( this.history ).length;
	}

	move ( d ) {
		if ( d & Direction.U ) --this.row;
		if ( d & Direction.D ) ++this.row;
		if ( d & Direction.L ) --this.col;
		if ( d & Direction.R ) ++this.col;
		this.track_history();
	}

	follow ( other ) {
		if ( Math.abs( this.row - other.row ) <= 1
		&&   Math.abs( this.col - other.col ) <= 1 )
			return;
		let d = 0;
		if ( this.row > other.row ) d |= Direction.U;
		if ( this.row < other.row ) d |= Direction.D;
		if ( this.col > other.col ) d |= Direction.L;
		if ( this.col < other.col ) d |= Direction.R;
		this.move( d );
	}
}

function range ( start, stop, step ) {
	var a = [ start ], b = start;
	while ( b < stop ) {
		a.push( b += step || 1 );
	}
	return a;
}

function main ( filename, knot_count, desc ) {
	if ( knot_count < 2 ) {
		console.error( "knot_count too low" );
		return;
	}

	let knots = [];
	for ( var i of range( 0, knot_count - 1 ) ) {
		knots[i] = new Knot( 0, 0 );
	}

	require( 'fs' ).readFile( filename, 'utf8', ( err, input ) => {
		if ( err ) {
			console.error( err );
			return;
		}
		input.split( /\r?\n/ ).forEach( line => {
			let parts = line.split( ' ' );
			let d = parts[0];
			let move_count = parseInt( parts[1], 10 );
			for ( var i of range( 1, move_count ) ) {
				knots[0].move( Direction[d] );
				for ( let ix of range( 1, knot_count - 1 ) ) {
					knots[ix].follow( knots[ix - 1] );
				}
			}
		} );
		console.log( `${desc}: ${ knots[knot_count - 1].history_size }` );
	} );
}

main( process.env.ADVENT_INPUT, 2, "PART1" );
main( process.env.ADVENT_INPUT, 10, "PART2" );

File: 2022/09/Python/main.py
#!/usr/bin/python3.10

import os

class Knot:
	def __init__ ( self, row, col ):
		self.row, self.col, self.history = row, col, {}
		self.track_history()

	def track_history ( self ):
		self.history[ "{},{}".format( self.row, self.col ) ] = 1

	def history_size ( self ):
		return len( self.history )

	def move ( self, direction ):
		if "U" in direction: self.row -= 1
		if "D" in direction: self.row += 1
		if "L" in direction: self.col -= 1
		if "R" in direction: self.col += 1
		self.track_history()

	def follow ( self, other ):
		if abs( self.row - other.row ) <= 1 and abs( self.col - other.col ) <= 1:
			return
		direction = '';
		if self.row > other.row: direction += 'U'
		if self.row < other.row: direction += 'D'
		if self.col > other.col: direction += 'L'
		if self.col < other.col: direction += 'R'
		self.move( direction )

def solve ( filename, knot_count, desc ):
	if knot_count <= 1:
		raise Exception( "knot_count too low" )
	knots = [ Knot( 0, 0 ) for _ in range( knot_count ) ]
	for line in open( filename, "r" ).readlines():
		direction, move_count = line.split( ' ' )
		for i in range( int( move_count ) ):
			knots[0].move( direction )
			for ix in range( 1, knot_count ):
				knots[ix].follow( knots[ix-1] );
	print( "%s: %d" % ( desc, knots[knot_count-1].history_size() ) )

solve( os.getenv( 'ADVENT_INPUT' ),  2, 'PART1' );
solve( os.getenv( 'ADVENT_INPUT' ), 10, 'PART2' );
